pkg/cache/memcached.go

```diff
@@ -143,8 +123,42 @@ func (c *Memcached) fetch(ctx context.Context, keys []string) (found []string, b
 	return
 }
 
+// FetchKey gets a single key from the cache
+func (c *Memcached) FetchKey(ctx context.Context, key string) ([]byte, bool) {
+	select {
+	case <-ctx.Done():
+		return nil, false
+	default:
+	}
+
+	const method = "Memcache.Get"
+	var item *memcache.Item
+	err := measureRequest(ctx, method, c.requestDuration, memcacheStatusCode, func(_ context.Context) error {
+		var err error
+		item, err = c.memcache.Get(key, memcache.WithAllocator(cacheBufAllocator))
+		if err != nil {
+			if errors.Is(err, memcache.ErrCacheMiss) {
+				level.Debug(c.logger).Log("msg", "Failed to get key from memcached", "err", err, "key", key)
+			} else {
+				level.Error(c.logger).Log("msg", "Error getting key from memcached", "err", err, "key", key)
+			}
+		}
+		return err
+	})
+	if err != nil {
+		return nil, false
+	}
+	return item.Value, true
+}
+
 // Store stores the key in the cache.
 func (c *Memcached) Store(ctx context.Context, keys []string, bufs [][]byte) {
+	select {
```

Comment:
Do you want to put this inside the loop?  Not sure how long the Store calls take generally, but checking at the top of the loop could possibly help.